File: 2020/day1/c++/part1.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <algorithm>

using namespace std;

int main()
{
    string s;
    vector<int> numbers;
    ifstream input_file;

    input_file.open("../input.txt", ios::in);
    if (!input_file) {
        cerr << "cannot open file" << endl;
        exit(1);
    }
    while (getline(input_file, s))
        numbers.push_back(stoi(s));
    input_file.close();

	if (numbers.empty())
		return 0;

    // when array is sorted, can use lower_bound, which is O(log n), instead of find, which is O(n)
    sort(numbers.begin(), numbers.end());

    for (auto i = numbers.begin(); i != numbers.end()-1; i++) {
        int sought = 2020 - *i;
        auto it = lower_bound(i+1, numbers.end(), sought);
        if (it != numbers.end() && *it == sought) {
            cout << *i * sought << endl;
            break; // comment if there could be more than one pair
        }
    }

    return 0;
}

File: 2020/day1/java/Part1.java
import java.io.File;
import java.io.FileNotFoundException;
import java.util.*;

class Part1
{
    private static List<Integer> readFile()
    {
        List<Integer> numbers = new ArrayList<Integer>();
        try {
            File inputFile = new File("../input.txt");
            Scanner inputScanner = new Scanner(inputFile);
            while (inputScanner.hasNextLine()) {
                String line = inputScanner.nextLine();
                int x = Integer.parseInt(line);
                numbers.add(x);
            }
            inputScanner.close();
        } catch (FileNotFoundException e) {
            System.out.println("Cannot open file");
        }
        return numbers;
    }

    public static void main(String[] args)
    {
        List<Integer> numbers = readFile();
        Collections.sort(numbers);

        //for (Integer i : numbers) {
        for (int i = 0 ; i < numbers.size(); i++) {
            int x = numbers.get(i);
            Integer sought = 2020 - x;
            int idx = numbers.indexOf(sought);
            if (idx > i && idx != -1) {
                System.out.println(x*numbers.get(idx));
            }
        }
    }
}
